File: IOS:iPadOS/Shared/WidgetSnapshot.swift
import Foundation

struct WidgetAuditItem: Codable, Identifiable, Hashable {
    let id: Int
    let tag: String
    let name: String
}

struct WidgetMaintenanceItem: Codable, Identifiable, Hashable {
    let id: Int
    let title: String
    let assetTag: String?
}

struct WidgetStockItem: Codable, Identifiable, Hashable {
    let id: String
    let name: String
    let remaining: Int
    let kindLabel: String
}

struct WidgetSnapshot: Codable {
    var isConfigured: Bool = false
    var serverHost: String?
    var savedAt: Date = .distantPast

    var auditsOverdue: Int = 0
    var auditsDueToday: Int = 0
    var auditsDueSoon: Int = 0
    var openMaintenance: Int = 0
    var totalAssets: Int = 0
    var deployedAssets: Int = 0
    var lowStockItems: Int = 0

    var topOverdueAudits: [WidgetAuditItem] = []
    var topOpenMaintenance: [WidgetMaintenanceItem] = []
    var topLowStockItems: [WidgetStockItem] = []

    static let empty = WidgetSnapshot()

    var hasData: Bool {
        isConfigured && totalAssets > 0
    }

    var formattedSavedAt: String {
        guard savedAt != .distantPast else { return "—" }
        let formatter = RelativeDateTimeFormatter()
        formatter.unitsStyle = .abbreviated
        return formatter.localizedString(for: savedAt, relativeTo: Date())
    }

    var availableAssets: Int {
        max(totalAssets - deployedAssets, 0)
    }

    var hasActionItems: Bool {
        !topOverdueAudits.isEmpty || !topOpenMaintenance.isEmpty || !topLowStockItems.isEmpty
    }
}

enum WidgetSnapshotStore {
    private static var containerURL: URL? {
        FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: WidgetConstants.appGroupID)
    }

    private static var fileURL: URL? {
        containerURL?.appendingPathComponent(WidgetConstants.snapshotFileName)
    }

    static func load() -> WidgetSnapshot {
        guard let url = fileURL,
              let data = try? Data(contentsOf: url),
              let snapshot = try? JSONDecoder().decode(WidgetSnapshot.self, from: data)
        else {
            return .empty
        }
        return snapshot
    }

    static func save(_ snapshot: WidgetSnapshot) {
        guard let url = fileURL,
              let data = try? JSONEncoder().encode(snapshot)
        else { return }
        try? data.write(to: url, options: .atomic)
    }

    static func clear() {
        guard let url = fileURL else { return }
        try? FileManager.default.removeItem(at: url)
    }
}

File: Android/app/src/main/java/com/callandt/snipemobile/widget/WidgetSnapshot.kt
package com.callandt.snipemobile.widget

import kotlinx.serialization.Serializable

@Serializable
data class WidgetAuditItem(
    val id: Int,
    val tag: String,
    val name: String,
)

@Serializable
data class WidgetMaintenanceItem(
    val id: Int,
    val title: String,
    val assetTag: String? = null,
)

@Serializable
data class WidgetStockItem(
    val id: String,
    val name: String,
    val remaining: Int,
    val kindLabel: String,
)

@Serializable
data class WidgetSnapshot(
    val isConfigured: Boolean = false,
    val serverHost: String? = null,
    val savedAtEpochMs: Long = 0L,
    val auditsOverdue: Int = 0,
    val auditsDueToday: Int = 0,
    val auditsDueSoon: Int = 0,
    val openMaintenance: Int = 0,
    val totalAssets: Int = 0,
    val deployedAssets: Int = 0,
    val lowStockItems: Int = 0,
    val topOverdueAudits: List<WidgetAuditItem> = emptyList(),
    val topOpenMaintenance: List<WidgetMaintenanceItem> = emptyList(),
    val topLowStockItems: List<WidgetStockItem> = emptyList(),
) {
    val hasData: Boolean get() = isConfigured && totalAssets > 0
    val availableAssets: Int get() = maxOf(totalAssets - deployedAssets, 0)
    val hasActionItems: Boolean
        get() = topOverdueAudits.isNotEmpty() ||
            topOpenMaintenance.isNotEmpty() ||
            topLowStockItems.isNotEmpty()

    companion object {
        val empty = WidgetSnapshot()
    }
}

/** Persists widget data to app-private storage. */
object WidgetSnapshotStore {
    private const val FILE_NAME = "widget_snapshot.json"

    fun load(context: android.content.Context): WidgetSnapshot {
        val file = context.filesDir.resolve(FILE_NAME)
        if (!file.exists()) return WidgetSnapshot.empty
        return runCatching {
            com.callandt.snipemobile.data.model.SnipeJson.decodeFromString(
                WidgetSnapshot.serializer(),
                file.readText(),
            )
        }.getOrDefault(WidgetSnapshot.empty)
    }

    fun save(context: android.content.Context, snapshot: WidgetSnapshot) {
        runCatching {
            val data = com.callandt.snipemobile.data.model.SnipeJson.encodeToString(snapshot)
            context.filesDir.resolve(FILE_NAME).writeText(data)
        }
    }

    fun clear(context: android.content.Context) {
        context.filesDir.resolve(FILE_NAME).delete()
    }
}
